index.js

```diff
@@ -145,72 +143,52 @@ ShareDbMongo.prototype._connect = function(mongo, options) {
+    var result = {
+      mongo: mongoClient,
+      mongoClient: mongoClient,
+      mongoPoll: mongoPollClient,
+      mongoPollClient: mongoPollClient
     };
-    async.parallel(tasks, function(err, results) {
-      if (err) throw err;
-      var mongoClient = results.mongoClient;
-      var mongoPollClient = results.mongoPollClient;
-      if (isLegacyMongoClient(mongoClient)) {
-        self.mongo = self._mongoClient = mongoClient;
-        self.mongoPoll = self._mongoPollClient = mongoPollClient;
-      } else {
-        self.mongo = mongoClient.db();
-        self._mongoClient = mongoClient;
-        self.mongoPoll = mongoPollClient.db();
-        self._mongoPollClient = mongoPollClient;
-      }
-      self._flushPendingConnect();
-    });
-    return;
-  }
-  var finish = function(err, client) {
-    if (err) throw err;
-    if (isLegacyMongoClient(client)) {
-      self.mongo = self._mongoClient = client;
-    } else {
-      self.mongo = client.db();
-      self._mongoClient = client;
+    if (!isLegacyMongoClient(mongoClient)) {
+      result.mongo = mongoClient.db();
+      if (mongoPollClient) result.mongoPoll = mongoPollClient.db();
     }
-    self._flushPendingConnect();
-  };
-  if (typeof mongo === 'function') {
-    mongo(finish);
-    return;
-  }
-  // TODO: Don't pass options directly to mongodb.connect();
-  // only pass options.mongoOptions
-  var mongoOptions = options.mongoOptions || options;
-  connect(mongo, mongoOptions)(finish);
+    return result;
+  });
 };
 
 function connect(mongo, options) {
-  if (typeof mongo === 'function') return mongo;
-  return function(callback) {
-    options = Object.assign({}, options);
-    delete options.mongo;
-    delete options.mongoPoll;
-    delete options.mongoPollOptions;
-    delete options.pollDelay;
-    delete options.disableIndexCreation;
-    delete options.allowAllQueries;
-    delete options.allowJSQueries;
-    delete options.allowAllQueries;
-    delete options.allowAggregateQueries;
-    delete options.getOpsWithoutStrictLinking;
-
-    if (typeof mongodb.connect === 'function') {
-      mongodb.connect(mongo, options, callback);
-    } else {
-      var client = new mongodb.MongoClient(mongo, options);
-      client.connect(callback);
-    }
-  };
+  if (typeof mongo === 'function') {
+    return new Promise(function(resolve, reject) {
+      mongo(function(error, db) {
+        if (error) return reject(error);
+        resolve(db);
+      });
+    });
+  }
+
+  options = Object.assign({}, options);
+  delete options.mongo;
+  delete options.mongoPoll;
+  delete options.mongoPollOptions;
+  delete options.pollDelay;
+  delete options.disableIndexCreation;
+  delete options.allowAllQueries;
+  delete options.allowJSQueries;
+  delete options.allowAllQueries;
+  delete options.allowAggregateQueries;
+  delete options.getOpsWithoutStrictLinking;
+
+  if (typeof mongodb.connect === 'function') return mongodb.connect(mongo, options);
```

Comment:
Minor - I think the if-else was clearer than this early-return style, since if-else indicates the result would be one or the other